Feature: ui grid filtering

jniles opened the pull request:
For full details of this change, see the commit messages.  The changes are best explained via screenshots:

**Screenshots**
![thepatientregistry](https://cloud.githubusercontent.com/assets/896472/16164821/27e0536e-34d8-11e6-8a1e-fff81211aaf4.png)
_Fig 1: The Unfiltered Patient Registry_

![filterform](https://cloud.githubusercontent.com/assets/896472/16164820/27dd8fd0-34d8-11e6-9c46-62a3fe4bf9ff.png)
_Fig 2: The New Filter Form_

![filteredregistry](https://cloud.githubusercontent.com/assets/896472/16164819/27d8414c-34d8-11e6-94ef-b1e0433b2da5.png)
_Fig 3: The New Filter Bar_

![filteredpdfreport](https://cloud.githubusercontent.com/assets/896472/16164818/27d2de50-34d8-11e6-81b7-cb205ab7e5ac.png)
_Fig 4: Filters appear on rendered PDF report_

**Major Features**
1. Clicking "x" on any filter will remove it and re-query the database.
2. The filters are cached, so that page refresh remembers the filters.
3. The filter form will be re-populated with filters from the grid when it is opened a second time.
4. The "Clear all Filters" button removes all filters from the grid.
5. The dates on the form have a clearer UI (start, end).
6. The PDF now shows what filters rendered it in a bar identical to the client's bar.

Closes #492.  Closes #336.

---

Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

jniles committed:
refactor(filters): improve Patient Registry filter

This commit refactors the Patient Registry filters to automatically
template into the registration page nice-looking filters.

Features landing in this commit:
 1. Filters are now cancellable, which refreshes the grid with the
 remaining filters.
 2. Filters are automatically templated into the PDF report generated by
 clicking the `print()` button on the registry page.
 3. "Clear All Filters" button clears all filters on the grid and
 refreshes the grid's data via an HTTP request
 4. Filters are cached so that a page refresh does not clear the
 filters.
 5. The filter form UI has drastically improved, allows start and end
 dates to be laid out side by side.

There are some future directions left to do after the commit.
 1. Improve rendering of filters in PDF reports.  The dates are not
 properly templated in, now do the values have much meaning.
 2. Rename the `bhFiltersApplied` directive to something a bit more
 descriptive of it's form and function.  Something like `bhFilterBar` or
 similar.
 3. Implement filtering tests as unit tests for search modal.

jniles committed:
feat(filters): improve filter ui rendering

This commit improves the filters ui again by introducing the following
upgrades:
 1. The server uses `moment` to parse dates in the date filters.  The
 code on the server for rendering the filters is very similar to the
 clientside codebase, which is unavoidable.  There might be an
 optimisation down the road to reduce this replication, but I think this
 is acceptable for now.

 2. The filters are now passed into the filter form and preset there.
 That way, when you go to filter a second time, all the data is
 preserved.

 3. A weird bug with caching filters was fixed.  The detailed flag no
 longer shows up as a potential filter, and refreshing the page after
 printing no longer breaks the browser.

 4. The date rendering has been improved in the handlebar's date helper
 function.  If an invalid date occurs, it will simply render the empty
 string.

 5. The `$uibModal` handler now implements the behavior described in
 #506.

Closes #492.  Closes #336.

sfount commented on client/test/e2e/patient/registry.search.js:
```diff
@@ -0,0 +1,172 @@
+    FU.buttons.submit();
+
+    expectNumberOfGridRows(1);
+    expectNumberOfFilters(1);
```
:+1:

sfount commented on client/src/js/components/bhFiltersApplied.js:
```diff
@@ -1,7 +1,29 @@
+
+bhFiltersAppliedController.$inject = [ '$scope', '$filter' ];
+
+function bhFiltersAppliedController($scope, $filter) {
```
Could you leave a quick comment explaining what this directive does? 

Just to make sure I have it straight 
- Displays all filters passed in through the filters attribute
- Runs any Angular filter through the $filter service before displaying the value

A very short reasoning behind the $watch may be useful for the next developer working on filters.

jniles replied:
Sure, I'll do that in a second and ping you back when it is complete.

jniles committed:
docs(bhFiltersApplied): improve directive docs

This commit adds doc comments to the bhFiltersApplied directive for
better readability and developer productivity in the future.
